synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -291,17 +284,30 @@ async def action_for_event_by_user(
         else:
             ignorers = frozenset()
 
+        users = list(rules_by_user.keys())
+        profiles = await self.store.get_users_in_room_with_profiles(
+            event.room_id, users
+        )
+
         for uid, rules in rules_by_user.items():
             if event.sender == uid:
                 continue
 
             if uid in ignorers:
                 continue
 
+            # This is a check for the case where user joins a room without being allowed to see history, and then the server
+            # receives a delayed event from before the user joined, which they should not be pushed for
+            visible = await filter_events_for_client_with_state(
+                self.store, uid, event, context
+            )
```

Comment:
Thinking about this some more, we probably want to pull this function call out of the loop and change it to work on a batch of users. This is a hot path and the `rules_by_user` dict could be huge, so hitting the DB one by one could cause it to take quite a lot of wall clock time.